### PR TITLE
Feature: noise integration

### PR DESCRIPTION
**Pull request by hJaffaliColibritd**

### CODE REVIEW

**Review by Henri-ColibrITD — CHANGES_REQUESTED** [review]

File: mpqp/noise/noise_model.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+                raise NotImplementedError(f"Depolarizing channel is not implemented in Braket for more than 2 qubits.")
+            elif self.dimension == 2:
+                return TwoQubitDepolarizing(probability=self.proba)
+            else:
+                return BraketDepolarizing(probability=self.proba)
+
+        elif language == Language.MY_QLM:
+            if self.dimension > 2:
+                raise NotImplementedError(f"Depolarizing channel is not implemented in the QLM for more than 2 qubits.")
+            elif self.dimension == 2 and self.gates is None:
```
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/noise/noise_model.py:249]: Indeed. We can also replace by ``if not self.gates`` right ?

File: mpqp/noise/noise_model.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+             [ 0.        +0.j -0.31622777+0.j]]
+            >>> type(qlm_depo)
+            qat.quops.quantum_channels.QuantumChannelKraus
+
+        """
+        if language == Language.BRAKET:
+            if self.dimension > 2:
+                raise NotImplementedError(f"Depolarizing channel is not implemented in Braket for more than 2 qubits.")
+            elif self.dimension == 2:
+                return TwoQubitDepolarizing(probability=self.proba)
```
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/noise/noise_model.py:240]: Modifier the attribute ``self.proba`` so it can only be of type float. Move the symbolic case to 3M-TODO

File: mpqp/execution/devices.py [hunk 3]
```diff
@@ -113,6 +127,34 @@ def is_gate_based(self) -> bool:
     def is_simulator(self) -> bool:
         return True
 
+    def is_noisy_simulator(self) -> bool:
```
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/execution/devices.py:161]: By definition all real devices are noisy, so we needed to define it only for simulators. One can still use the conjonction of is_simulator() and is_noisy() to check if the device is a noisy simulator, but this will made us call two functions, and require to define is_noisy so it returns also True for real devices

File: mpqp/core/instruction/gates/native_gates.py [hunk 4]
```diff
@@ -400,14 +440,17 @@ def to_other_language(
...
+            if qiskit_parameters is None:
+                qiskit_parameters = set()
             return UGate(
                 theta=_qiskit_parameter_adder(self.theta, qiskit_parameters),
                 phi=_qiskit_parameter_adder(self.phi, qiskit_parameters),
                 lam=_qiskit_parameter_adder(self.gamma, qiskit_parameters),
             )
+        elif language == Language.BRAKET:
+            return gates.U(self.theta, self.phi, self.gamma)
+            # 3M-TODO handle symbolic parameters
```
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/core/instruction/gates/native_gates.py:535]: I would let them as 3M-TODO, and maybe return an error if one of the parameters is Expr.

File: mpqp/core/instruction/gates/native_gates.py [hunk 5]
```diff
@@ -142,12 +143,15 @@ def to_other_language(
         theta = float(self.theta) if self._numeric_parameters else self.theta
         if language == Language.QISKIT:
             return self.qiskit_gate(_qiskit_parameter_adder(theta, qiskit_parameters))
+        elif language == Language.BRAKET:
+            # 3M-TODO: handle symbolic parameters for Braket
+            return self.braket_gate(theta)
```
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/core/instruction/gates/native_gates.py:138]: I'll add a task on Trello

File: mpqp/execution/providers/atos.py [hunk 6]
```diff
@@ -43,50 +86,88 @@ def get_local_qpu(device: ATOSDevice) -> QPUHandler:
...
+            from qlmaas.qpus import LinAlg  # type: ignore
+            return LinAlg()
+        elif device == ATOSDevice.QLM_MPS:
+            get_QLMaaSConnection()
+            from qlmaas.qpus import MPS  # type: ignore
+            return MPS()
+        elif device == ATOSDevice.QLM_NOISYQPROC:
+            get_QLMaaSConnection()
+            from qlmaas.qpus import NoisyQProc  # type: ignore
+            qpu = NoisyQProc(sim_method="stochastic", n_samples=job.measure.shots if job.measure is not None else 0)
```
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/execution/providers/atos.py]: I removed the possibility of job to be None, we never use it with a None argument.

File: tests/noise/test_noisy_execution.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            # CRk(4, 2, 0)
+        ]
+    )
+
+
+@pytest.fixture
+def devices():
+    devices: list[AvailableDevice] = [AWSDevice.BRAKET_LOCAL_SIMULATOR]
+    if "-l" in sys.argv or "--long" in sys.argv:
+        devices.append(ATOSDevice.QLM_NOISYQPROC)
```
Comment (hJaffaliColibritd) [inline tests/noise/test_noisy_execution.py:45]: We lost after merging/refacto here the device QLM_MPO (that is a noisy simu also) when the job is of type SAMPLE, we need to test it too

File: tests/noise/test_noisy_execution.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            CZ(1, 2),
+            # CRk(4, 2, 0)
+        ]
+    )
+
+
+@pytest.fixture
+def devices():
+    devices: list[AvailableDevice] = [AWSDevice.BRAKET_LOCAL_SIMULATOR]
+    if "-l" in sys.argv or "--long" in sys.argv:
```
Comment (hJaffaliColibritd) [inline tests/noise/test_noisy_execution.py:44]: This works for this test, but no more for other tests in tests/execution/providers/atos.py for instance

File: mpqp/execution/providers/aws.py [hunk 9]
```diff
@@ -22,29 +24,103 @@
...
+    for instr in braket_circuit.instructions:
+        if isinstance(instr.operator, BraketMeasure):
+            stored_measurements.append(instr)
+        else:
+            other_instructions.append(instr)
+
+    noisy_circuit = Circuit(other_instructions)
+
+    for noise in noises:
+        if set(noise.targets) == set(range(nb_qubits)):
```
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/execution/providers/aws.py:65]: We can improve, but it seems that we will always end up with 4 different cases, involving 2 nested if/else

File: mpqp/execution/providers/atos.py [hunk 10]
```diff
@@ -1,93 +1,219 @@
...
-    Returns:
-        A myQLM Job and the right myQLM QPUHandler on which it will be submitted.
+    Raises:
+        ValueError: If the required backend is a local simulator.
     """
+    if not device.is_remote():
+        raise ValueError(
+            f"Excepted a remote device, but got a local myQLM simulator {device}"
+        )
+    if len(job.circuit.noises) > 0:
```
Comment (Henri-ColibrITD) [inline mpqp/execution/providers/atos.py:125]: what if `job` is `None` ?

File: mpqp/execution/providers/aws.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Henri-ColibrITD) [reply mpqp/execution/providers/aws.py]: I pushed a suggested modification, tell me what you think about it

File: mpqp/execution/providers/atos.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Henri-ColibrITD) [reply mpqp/execution/providers/atos.py]: I still see it as `Optional`

File: mpqp/noise/noise_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Henri-ColibrITD) [reply mpqp/noise/noise_model.py]: we could, but in these cases I prefer being explicit rather than relying on truthy values. This is one of the cases where I find that writing more is beneficial, because it allows you to see that you are testing specifically for an empty list (and not a `None` value, a `0`, or any of these other falsy values)

File: mpqp/execution/providers/aws.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/execution/providers/aws.py]: Agreed

File: mpqp/execution/devices.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hJaffaliColibritd) [reply mpqp/execution/devices.py:161]: I said real devices are noisy, meaning quantum computers. The question of being noisy or not is only relevant for simulators, that is why I think we should keep the restriction on the name

File: mpqp/core/circuit.py [hunk 11]
```diff
@@ -661,18 +749,35 @@ def to_other_language(
...
             <class 'qiskit.circuit.quantumcircuit.QuantumCircuit'>
+            >>> circuit2 = QCircuit([H(0), CZ(0,1), Depolarizing(0.6, [0])])
+            >>> braket_circuit = circuit2.to_other_language(Language.BRAKET)
+            >>> print(braket_circuit)  # doctest: +NORMALIZE_WHITESPACE
+            T  : │         0         │         1         │
+                  ┌───┐ ┌───────────┐       ┌───────────┐
+            q0 : ─┤ H ├─┤ DEPO(0.6) ├───●───┤ DEPO(0.6) ├─
+                  └───┘ └───────────┘   │   └───────────┘
+                                      ┌─┴─┐
+            q1 : ─────────────────────┤ Z ├───────────────
```
Comment (Henri-ColibrITD) [inline mpqp/core/circuit.py:778]: the to_other_language for the noise seem wrong: the noise is applied to all qubits here but it shouldn't be the case

File: mpqp/execution/devices.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Henri-ColibrITD) [reply mpqp/execution/devices.py:161]: Oh, right, I can't read x)
ok, you convinced me

File: mpqp/core/circuit.py (hunk 11 above)
Comment (hJaffaliColibritd) [reply mpqp/core/circuit.py:778]: Solved

**Review by Henri-ColibrITD — APPROVED** [review]
